control-plane/p4RuntimeSerializer.cpp

```diff
@@ -661,9 +661,26 @@ getMatchFields(const IR::P4Table* table, ReferenceMap* refMap, TypeMap* typeMap)
           typeMap->getType(keyElement->expression->getNode(), true);
         BUG_CHECK(matchFieldType != nullptr,
                   "Couldn't determine type for key element %1%", keyElement);
-
+        size_t w = 0;
+		// Table key needs different newtype handling because the top newtype
+		// could be Type_Bits.
+        if (matchFieldType->is<IR::Type_Newtype>()) {
```

Comment:
I'm confused by this code. First, I would imagine that the `width_bits` method should work on custom "new" types as well; if this is not the case, maybe the `width_bits` implementation should be fixed?
Secondly, I may be wrong, but it looks to me like you are only looking 2 levels deep, but what about this case:
```
type bit<9> typeA;
type typeA typeB;
type typeB typeC;
...
```
Shouldn't you be using a while loop or a recursive function. BTW I believe `width_bits` should work because it should call `width_bits` recursively on the underlying type...
Finally, you do not seem to be populated the `type_name` field in the `MatchField` P4Info message.

Reply:
I have tested 3 levels of type in p416-type-use3.p4.  The Type_NewType is an interesting beast.  If one walks Type_NewType one ends up with Type_Name which cannot be used to compute width_bits.   p4c does support width_bits, but does not use recursion.  Let me look into a universal width_bits() for Type_Newtype in type.def and get back.

Reply:
Looking at type.def, `Type_Newtype` has a `width_bits` method. If this method doesn't work for some cases like the one below, we need to fix it:
```
type bit<9> typeA;
type typeA typeB;
```

By the way @mbudiu-vmw , is it correct that calling `getTypeType` on a IR node of type `Type_Newtype` will return "the most underlying type". In other words, if I call `getTypeType` on `typeB` in the example above, it will return an object of type `Type_Bits`? Hemant's code seems to be assuming this.

Reply:
@antoninbas A wrinkle here -- If the user does "stack" `type` definitions like this, should P4Runtime use the topmost one in the P4info file?  Or the bottom-most one?  I suspect the top-most one would be the right answer, although as mentioned elsewhere, I doubt user programs will 'stack' such `type` definitions very often.  Still nice to have a good deterministic answer for what the compiler and other tools do in such a situation, though.

Reply:
The type in the P4Info should be the type of the expression

Reply:
So based on Mihai's comment, this code should work in all cases and is much simpler:
```
// can be Type_Newtype
auto* matchFieldType = typeMap->getType(keyElement->expression->getNode(), true);
// cannot be Type_Newtype
auto* canonicalType = typeMap->getTypeType(matchFieldType);
// check that canonicalType is a supported type (Type_Bits)
int width = canonicalType->width_bits();
```
@mbudiu-vmw please let us know if I'm wrong

Reply:
@antoninbas First, I am using a deeply nested type and typedef program provided by Andy: `p416-type-use3.p4`

Using your code above, the problem is with the getTypeType function.  If given a newtype, the function just returns the newtype.  I have narrowed down the problem to this line of code.

https://github.com/p4lang/p4c/blob/master/frontends/p4/typeMap.cpp#L95

Need to investigate Type_Type(). 

canonicalType that failed width_bits() is included.  First, see the failure below. This is why I have the code that way it is.  From newtype, get the type, which is Type_Name and then use getTypeType().  

```
Error compiling
terminate called after throwing an instance of 'Util::CompilerBug'
  what():  In file: /tmp/p4c/ir/type.def:335
Compiler Bug: Type_Name is not a canonical type, use getTypeType()?
 ```

```
[1101] Type_Newtype name=CustomDT_t declid=66
  annotations: [1099] Annotations
  type: [1098] Type_Name
    path: [1097] Path name=CustomD_t absolute=0
```

Reply:
This is the scoop.  With my checked in code for `getMatchFields()`, `getTypeType()` is sent a `Type_Name` and returns a `Type_Newtype`.  Note, both types have the same `name` and this is why in my prior note, I got fooled saying the same type is returned.  `getTypeType()` returns the penultimate type.  The example below has CustomTT_t and `getTypeType()` returns with an entity whose type is CustomT_t.  Thus, we do not need any while loop.

```
BEFORE call to getTypeType()
[1148] Type_Name
  path: [1147] Path name=CustomTT_t absolute=0
AFTER
[1111] Type_Newtype name=CustomTT_t declid=68
  annotations: [1109] Annotations
  type: [1108] Type_Name
    path: [1107] Path name=CustomT_t absolute=0
```